src/backend/distributed/utils/resource_lock.c

```diff
@@ -534,12 +573,21 @@ SerializeNonCommutativeWrites(List *shardIntervalList, LOCKMODE lockMode)
 	ShardInterval *firstShardInterval = (ShardInterval *) linitial(shardIntervalList);
 	int64 firstShardId = firstShardInterval->shardId;
 
-	if (ReferenceTableShardId(firstShardId) && ClusterHasKnownMetadataWorkers() &&
-		!IsFirstWorkerNode())
+	if (ReferenceTableShardId(firstShardId))
 	{
-		LockShardListResourcesOnFirstWorker(lockMode, shardIntervalList);
+		if (ClusterHasKnownMetadataWorkers() && !IsFirstWorkerNode())
+		{
+			LockShardListResourcesOnFirstWorker(lockMode, shardIntervalList);
```

Comment:
are we doing this twice now? (again in LockReferencedReferenceShardResources)

Reply:
> are we doing this twice now? (again in LockReferencedReferenceShardResources)

Here it is for the relation itself, in `LockReferencedReferenceShardResources` it is for referenced relations. I am changing the variable names there to avoid confusion.